pandas/_libs/algos.pyx

```diff
@@ -793,216 +796,280 @@ ctypedef fused rank_t:
     else:
-        values = np.asarray(in_arr).copy()
-
-    keep_na = na_option == 'keep'
+        masked_vals = in_arr.copy()
 
     if rank_t is object:
-        mask = missing.isnaobj(values)
-    elif rank_t is float64_t:
-        mask = np.isnan(values)
+        mask = missing.isnaobj(masked_vals)
     elif rank_t is int64_t:
-        mask = values == NPY_NAT
-
-    # double sort first by mask and then by values to ensure nan values are
-    # either at the beginning or the end. mask/(~mask) controls padding at
-    # tail or the head
-    if rank_t is not uint64_t:
-        if ascending ^ (na_option == 'top'):
-            if rank_t is object:
-                nan_value = Infinity()
-            elif rank_t is float64_t:
-                nan_value = np.inf
-            elif rank_t is int64_t:
-                nan_value = np.iinfo(np.int64).max
+        mask = (masked_vals == NPY_NAT).astype(np.uint8)
+    elif rank_t is float64_t:
+        mask = np.isnan(masked_vals).astype(np.uint8)
+    else:
+        mask = np.zeros(shape=len(masked_vals), dtype=np.uint8)
 
-            order = (values, mask)
+    if ascending ^ (na_option == 'top'):
+        if rank_t is object:
+            nan_fill_val = Infinity()
+        elif rank_t is int64_t:
+            nan_fill_val = np.iinfo(np.int64).max
+        elif rank_t is uint64_t:
+            nan_fill_val = np.iinfo(np.uint64).max
         else:
-            if rank_t is object:
-                nan_value = NegInfinity()
-            elif rank_t is float64_t:
-                nan_value = -np.inf
-            elif rank_t is int64_t:
-                nan_value = np.iinfo(np.int64).min
-
-            order = (values, ~mask)
-        np.putmask(values, mask, nan_value)
+            nan_fill_val = np.inf
+        order = (masked_vals, mask, labels)
     else:
-        mask = np.zeros(shape=len(values), dtype=bool)
-        order = (values, mask)
+        if rank_t is object:
+            nan_fill_val = NegInfinity()
+        elif rank_t is int64_t:
+            nan_fill_val = np.iinfo(np.int64).min
+        elif rank_t is uint64_t:
+            nan_fill_val = 0
+        else:
+            nan_fill_val = -np.inf
 
-    n = len(values)
-    ranks = np.empty(n, dtype='f8')
+        order = (masked_vals, ~mask, labels)
 
-    if rank_t is object:
-        _as = np.lexsort(keys=order)
-    else:
-        if tiebreak == TIEBREAK_FIRST:
-            # need to use a stable sort here
-            _as = np.lexsort(keys=order)
-            if not ascending:
-                tiebreak = TIEBREAK_FIRST_DESCENDING
-        else:
-            _as = np.lexsort(keys=order)
+    np.putmask(masked_vals, mask, nan_fill_val)
+
+    # lexsort using labels, then mask, then actual values
+    # each label corresponds to a different group value,
+    # the mask helps you differentiate missing values before
+    # performing sort on the actual values
+    _as = np.lexsort(order).astype(np.int64, copy=False)
```

Comment:
can you rename _as to something more readable, maybe lexsort_indexer

Reply:
Done